config/accounts.go

```diff
@@ -222,3 +225,40 @@ func (a *AccountIntegration) GetByIntegrationType(integrationType IntegrationTyp
 
 	return integrationEnabled
 }
+
+type AlternateBidderCodes struct {
+	Enabled  bool                                                   `mapstructure:"enabled" json:"enabled"`
+	Adapters map[openrtb_ext.BidderName]AdapterAlternateBidderCodes `mapstructure:"adapters" json:"adapters"`
+}
+
+type AdapterAlternateBidderCodes struct {
+	Enabled            bool                     `mapstructure:"enabled" json:"enabled"`
+	AllowedBidderCodes []openrtb_ext.BidderName `mapstructure:"allowedbiddercodes" json:"allowedbiddercodes"`
+}
+
+func (account *AlternateBidderCodes) IsValidBidderCode(bidder, alternateBidder openrtb_ext.BidderName) (bool, error) {
+	if alternateBidder == "" || bidder == alternateBidder {
+		return true, nil
+	}
+
+	if account.Adapters == nil {
+		return account.Enabled, nil
+	}
+
+	adapterCfg, ok := account.Adapters[bidder]
+	if !((ok && adapterCfg.Enabled) || (!ok && account.Enabled)) {
+		return false, nil
+	}
+
+	if len(adapterCfg.AllowedBidderCodes) == 0 || adapterCfg.AllowedBidderCodes[0] == "*" {
```

Comment:
Continuing discussion here w.r.t quoted comment at https://github.com/PubMatic-OpenWrap/prebid-server/pull/309#discussion_r888141638
> If we assume that an empty AllowedBidderCodes set means all bidder codes are allowed, which is a change from today's behavior. This requires PBS hosts to make a positive entry for every adapter to disable the new behavior, and continue creating a new entry every time a new adapter is released/enabled. I think we want to require a positive change for each adapter, like setting "*", as enabling this functionality requires additional trust in the behavior of an adapter that was not needed before. Or possibly allowing a default allowBidderCodes that determines what to do with an empty map.

I allowed `len(adapterCfg.AllowedBidderCodes) == 0` based on description on https://github.com/prebid/prebid-server/issues/2174. Quote:

> If true, then check account.alternateBidderCodes.adapters.BIDDER.allowedBidderCodes.
> - If doesn't exist or ["*"], then allow.

I'll update this code, making `AllowedBidderCodes` mandatory to be either `"*"` or a a list of bidders. Sounds good?

Reply:
Making it mandatory would not be good unless you also put in a default of only allowing each adapter to set itself as the seat (today's behavior) We don't want to break everyone's config when they upgrade, which is what would happen if a new parameter is added that is also mandatory. As far as I know, we only did that once, with the GDPR default setting, after much discussion concluding there was no safe default we could use without risking legal issues.

Reply:
What I meant was:
1. We use `account.alternateBidderCodes.enabled: true/false` as overall toggle for this feature.  (default would be `false` to keep things running as is after upgrade)
     - If `false`, reject the alternate bidder.
     - If `true`, check  `account.alternateBidderCodes.adapters.BIDDER.allowedBidderCodes`. Here,
         - If `allowedBidderCodes` is "*", allow the alternate bidder.
         - Or, if alternate bidder is in `allowedBidderCodes` list, allow the alternate bidder.
         - Else, if alternate bidder is not in `allowedBidderCodes` list, reject the alternate bidder.

2. With this, I don't see the use of `account.alternateBidderCodes.adapters.BIDDER.enabled: true/false`. Since, default empty `allowedBidderCodes` list would act as `account.alternateBidderCodes.adapters.BIDDER.enabled = false`.


Can we please have the description at https://github.com/prebid/prebid-server/issues/2174 updated with correct expectations?